### PR TITLE
Preserve comments on parameters in call expressions

### PR DESCRIPTION
**Pull request by yuit**
Fixing #3844

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -6925,6 +6956,18 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 emitComments(currentSourceFile, writer, trailingComments, /*trailingSeparator*/ false, newLine, writeComment);
             }
 
+            /**
+             * Emit trailing comments at the position. The term trailing comment is used here to describe following comment:
+             *      x, /comment1/ y
+             *        ^ => pos; the function will emit "comment1" in the emitJS
+             */
+            function emitTrailingCommentsOfPosition(pos: number) {
```
Comment (yuit) [inline src/compiler/emitter.ts]: I am not sure what would be an appropriate name here. So I wold love some suggestion

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -832,7 +832,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                             write(", ");
                         }
                     }
-                    emitNode(nodes[start + i]);
+                    let node = nodes[start + i];
+                    // This emitting is to make sure we emit following comment properly
+                    //   ...(x, /comment1/ y)...
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: You forgot the asterisks

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -832,7 +832,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
                         }
                     }
-                    emitNode(nodes[start + i]);
+                    let node = nodes[start + i];
+                    // This emitting is to make sure we emit following comment properly
+                    //   ...(x, /comment1/ y)...
+                    //         ^ => node.pos
+                    // "comment1" is not considered leading comment for node-y but rather
+                    // considered as trailing comment of the previous node.
+                    emitTrailingCommentsOfPosition(node.pos);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Your example is of a leading comment, here you're emitting trailing comments

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -832,7 +832,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                             write(", ");
                         }
                     }
-                    emitNode(nodes[start + i]);
+                    let node = nodes[start + i];
+                    // This emitting is to make sure we emit following comment properly
+                    //   ...(x, /comment1/ y)...
+                    //         ^ => node.pos
+                    // "comment1" is not considered leading comment for node-y but rather
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: just write 'y'

File: src/compiler/emitter.ts [hunk 5]
```diff
@@ -832,7 +832,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                             write(", ");
                         }
                     }
-                    emitNode(nodes[start + i]);
+                    let node = nodes[start + i];
+                    // This emitting is to make sure we emit following comment properly
+                    //   ...(x, /comment1/ y)...
+                    //         ^ => node.pos
+                    // "comment1" is not considered leading comment for node-y but rather
+                    // considered as trailing comment of the previous node.
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Why is that a trailing comment? That sounds broken

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -1976,6 +1983,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
             function emitPropertyAssignment(node: PropertyDeclaration) {
                 emit(node.name);
                 write(": ");
+                // This is to ensure that we emit comment in the following case:
+                //      For example:
+                //          obj = {
+                //              id: /comment1/ ()=>void
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Asterisks

**Comment by mhegazy** [comment]
:+1:

File: tests/baselines/reference/APISample_linter.js [hunk 7]
```diff
@@ -112,7 +112,7 @@ exports.delint = delint;
 var fileNames = process.argv.slice(2);
 fileNames.forEach(function (fileName) {
     // Parse a file
-    var sourceFile = ts.createSourceFile(fileName, readFileSync(fileName).toString(), 2 /* ES6 */, true);
+    var sourceFile = ts.createSourceFile(fileName, readFileSync(fileName).toString(), 2 /* ES6 */, /*setParentNodes */ true);
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/APISample_linter.js]: nice

File: src/compiler/emitter.ts [hunk 8]
```diff
@@ -6925,6 +6956,18 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 emitComments(currentSourceFile, writer, trailingComments, /*trailingSeparator*/ false, newLine, writeComment);
             }
 
+            /**
+             * Emit trailing comments at the position. The term trailing comment is used here to describe following comment:
+             *      x, /comment1/ y
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Asterisks

File: src/compiler/emitter.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This is perfect

**Pull request merged**